README.md
openmediavault-greyhole
=======================

OpenMediaVault plugin for Greyhole


debian/changelog
openmediavault-greyhole (3.1) stable; urgency=low

  * Changed dependency of OMV Package to = 3.0.13 to avoid accidental user updates

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Tue, 05 Apr 2016 16:06:52 -0400

openmediavault-greyhole (3.0) stable; urgency=low

  * Update for OMV 3.x

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Wed, 27 Jan 2016 12:18:04 -0600

openmediavault-greyhole (2.1) stable; urgency=low

  * Fix installation issue

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Mon, 06 Jul 2015 12:51:04 -0500

openmediavault-greyhole (2.0) stable; urgency=low

  * Update for OMV 2.x
  * Update locales

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Wed, 01 Jul 2015 14:31:53 -0500

openmediavault-greyhole (1.0.3) stable; urgency=low

  * Add custom debian/control field XB-Plugin-Section
  * Update locales

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Fri, 12 Dec 2014 06:23:09 -0600

openmediavault-greyhole (1.0.2) stable; urgency=low

  * Change log format

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Fri, 01 Aug 2014 06:40:04 -0500

openmediavault-greyhole (1.0.1) stable; urgency=low

  * Fix logging
  * Update locales

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Thu, 31 Jul 2014 17:37:28 -0500

openmediavault-greyhole (1.0) stable; urgency=low

  * Minor changes to info page
  * Update locales

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Fri, 18 Jul 2014 09:11:17 -0500

openmediavault-greyhole (0.6.3) stable; urgency=low

  * Modify package trigger handling
  * Update locales

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Mon, 14 Apr 2014 07:23:26 -0500

openmediavault-greyhole (0.6.2) stable; urgency=low

  * Fix service indicator status
  * Update locales

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Mon, 07 Apr 2014 06:47:26 -0500

openmediavault-greyhole (0.6.1) stable; urgency=low

  * Verify samba share exists before adding to greyhole config
  * Update locales
  * Update to run on sardaukar and kralizec

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Tue, 18 Mar 2014 13:27:26 -0500

openmediavault-greyhole (0.5.9) stable; urgency=low

  * Add samba share notify
  * Allow samba share edit

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Fri, 14 Feb 2014 13:03:26 -0500

openmediavault-greyhole (0.6.0) stable; urgency=low

  * Update to run on kralizec
  * Update locales

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Fri, 12 Feb 2014 07:29:26 -0500

openmediavault-greyhole (0.5.8) stable; urgency=low

  * Add ignored files/folders tab
  * Add info tab
  * Update locales

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Thu, 23 Jan 2014 07:17:26 -0500

openmediavault-greyhole (0.5.7) stable; urgency=low

  * Add option to make trash share writeable
  * Fix trash share samba share name
  * Update locales

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Wed, 22 Jan 2014 07:33:26 -0500

openmediavault-greyhole (0.5.6) stable; urgency=low

  * Add Trash Share option
  * Update locales

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Mon, 20 Jan 2014 10:45:26 -0500

openmediavault-greyhole (0.5.5) stable; urgency=low

  * Change Use Trash option on samba share
  * Hide database settings when enabled
  * Update copyrights
  * Update locales

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Sun, 19 Jan 2014 10:28:26 -0500

openmediavault-greyhole (0.5.4) stable; urgency=low

  * More fixes

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Sun, 08 Dec 2013 13:44:26 -0500

openmediavault-greyhole (0.5.3) stable; urgency=low

  * Add helper file

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Sat, 07 Dec 2013 10:41:26 -0500

openmediavault-greyhole (0.5.2) stable; urgency=low

  * Bug fixes

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Sat, 07 Dec 2013 09:22:26 -0500

openmediavault-greyhole (0.5.1) stable; urgency=low

  * Bug fixes

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Fri, 06 Dec 2013 14:36:26 -0500

openmediavault-greyhole (0.5) stable; urgency=low

  * Initial port to OMV 0.5

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Fri, 06 Dec 2013 11:16:26 -0500

openmediavault-greyhole (0.4) stable; urgency=low

  * Upgrade to xmlstarlet 1.3.1.

 -- Marcel Beck <marcel.beck@mbeck.org>  Sun, 09 Sep 2012 10:27:26 +0200

openmediavault-greyhole (0.3.3) unstable; urgency=high

  * Github Issue Fix #34 - [Mount shares locally] Add cifs-utils package + Tuning
  * Github Issue FIx #35 - [Mount shares locally] Password field isn't hidden

 -- Marcel Beck <marcel.beck@mbeck.org>  Sun, 10 Jun 2012 18:09:35 +0200

openmediavault-greyhole (0.3.2) unstable; urgency=high

  * Github Issue Fix #33 - [SMB Shares] Remove button isn't greyed when no line is selected
  * Github Issue Fix #32 - [Translation] Settings tab not translated
  * Github Issue Fix #31 - [Pool] Pool management error when using Replace option
  * Github Issue Fix #4 - Remove Samba Shares to the Greyhole Storage
  * Add script to mount samba shares locally for local use of greyhole
  * Update locale

 -- Marcel Beck <marcel.beck@mbeck.org>  Thu, 07 Jun 2012 19:47:48 +0200

openmediavault-greyhole (0.3.1) unstable; urgency=high

  * Various Fixes

 -- Marcel Beck <marcel.beck@mbeck.org>  Wed, 23 May 2012 18:23:42 +0200

openmediavault-greyhole (0.3) unstable; urgency=high

  * Github Issue Fix #21 - OMV 0.3 Port
  * Github Issue Fix #29 - Error when selecting log entry in OMV log tab

 -- Marcel Beck <marcel.beck@mbeck.org>  Fri, 20 Apr 2012 23:43:11 +0200

openmediavault-greyhole (0.2.2) stable; urgency=high

  * Github Issue #22 - Can't validate SMB dialog box when editing for sticky/trash options


var/www/openmediavault/js/omv/module/admin/service/greyhole/smb (2).js
/**
 * vim: tabstop=4
 *
 * @license http://www.gnu.org/licenses/gpl.html GPL Version 3
 * @author Stephane Bocquet <stephane_bocquet@hotmail.com>
 * @author Marcel Beck <marcel.beck@mbeck.org>
 * @copyright Copyright (c) 2011 Stephane Bocquet
 * @copyright Copyright (c) 2011 Marcel Beck
 * @version $Id: greyhole.js 12 2011-11-07 18:52:10Z
 *          stephane_bocquet@hotmail.com $
 *
 * This file is free software: you can redistribute it and/or modify it under
 * the terms of the GNU General Public License as published by the Free Software
 * Foundation, either version 3 of the License, or any later version.
 *
 * This file is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with
 * this file. If not, see <http://www.gnu.org/licenses/>.
 */
// require("js/omv/NavigationPanel.js")
// require("js/omv/data/DataProxy.js")
// require("js/omv/FormPanelExt.js")
// require("js/omv/form/SharedFolderComboBox.js")
// require("js/omv/form/plugins/FieldInfo.js")

// require("js/omv/module/greyhole/panel/navigation.js")

// require("js/omv/module/greyhole/admin/dialog/smb.js")

Ext.ns("OMV.Module.Storage.Greyhole.Admin");

/**
 * @class OMV.Module.Storage.Greyhole.Admin.SMBPanel
 * @derived OMV.grid.TBarGridPanel
 * Display list of configured filesystems.
 */
OMV.Module.Storage.Greyhole.Admin.SMBPanel = function (config) {
	var initialConfig = {
		hidePagingToolbar:false,
		autoReload       :true,
		stateId          :"85f1cbf2-23d3-4960-a803-b7fc34d42235",
		colModel         :new Ext.grid.ColumnModel({
			columns:[
				{
					header   :_("SMB Share"),
					sortable :true,
					dataIndex:"name",
					id       :"name"
				},
				{
					header   :_("Comment"),
					sortable :true,
					dataIndex:"comment",
					id       :"comment"
				},
				{
					header   :_("Files copies"),
					sortable :true,
					dataIndex:"num_copies",
					id       :"num_copies",
					width    :20
				},
				{
					header   :_("Sticky files"),
					sortable :true,
					dataIndex:"sticky_files",
					id       :"sticky_files",
					renderer :OMV.util.Format.booleanRenderer(),
					width    :20
				},
				{
					header   :_("Use Trash"),
					sortable :true,
					dataIndex:"trash",
					id       :"trash",
					renderer :OMV.util.Format.booleanRenderer(),
					width    :20
				}
			]
		})
	};
	Ext.apply(initialConfig, config);
	OMV.Module.Storage.Greyhole.Admin.SMBPanel.superclass.constructor.call(this, initialConfig);
};
Ext.extend(OMV.Module.Storage.Greyhole.Admin.SMBPanel, OMV.grid.TBarGridPanel, {
	initComponent:function () {
		this.store = new OMV.data.Store({
			autoLoad  :true,
			remoteSort:false,
			proxy     :new OMV.data.DataProxy({
				"service":"Greyhole",
				"method" :"getSMBList"
			}),
			reader    :new Ext.data.JsonReader({
				idProperty   :"uuid",
				totalProperty:"total",
				root         :"data",
				fields       :[
					{ name:"uuid" },
					{ name:"name" },
					{ name:"comment" },
					{ name:"num_copies" },
					{ name:"sticky_files" },
					{ name:"trash" }
				]
			})
		});
		OMV.Module.Storage.Greyhole.Admin.SMBPanel.superclass.initComponent.apply(this, arguments);
		// Register event handler
		// Reselect previous selected rows after the store has been
		// reloaded, e.g. to make sure toolbar is updated depending on
		// the latest row record values.
		this.getSelectionModel().previousSelections = [];
		this.store.on("beforeload", function (store, options) {
			var sm = this.getSelectionModel();
			var records = sm.getSelections();
			sm.previousSelections = [];
			Ext.each(records, function (record, index) {
				sm.previousSelections.push(record.get("uuid"));
			}, this);
		}, this);
		this.store.on("load", function (store, records, options) {
			var sm = this.getSelectionModel();
			var rows = [];
			if (Ext.isDefined(sm.previousSelections)) {
				for (var i = 0; i < sm.previousSelections.length; i++) {
					var index = store.findExact("uuid",
									sm.previousSelections[i]);
					if (index !== -1) {
						rows.push(index);
					}
				}
			}
			if (rows.length > 0) {
				sm.selectRows(rows);
			}
		}, this);
	},

	initToolbar:function () {
		var tbar = OMV.Module.Storage.Greyhole.Admin.SMBPanel.superclass.initToolbar.apply(this);

		tbar.remove(2);

		tbar.insert(2, {
			id      :this.getId() + "-remove",
			xtype   :"button",
			text    :_("Remove"),
			icon    :"images/delete.png",
			handler :this.cbRemoveShareBtnHdl,
			scope   :this,
			disabled:true
		});

		tbar.insert(3, {
			id     :this.getId() + "-fsck",
			xtype  :"button",
			text   :_("Files check"),
			icon   :"images/greyhole-fsck.png",
			handler:this.cbfsckBtnHdl,
			scope  :this
		});

		tbar.insert(4, {
			id     :this.getId() + "-unfsck",
			xtype  :"button",
			text   :_("Cancel all checks"),
			icon   :"images/greyhole-unfsck.png",
			handler:this.cbunfsckBtnHdl,
			scope  :this
		});

		return tbar;
	},

	cbSelectionChangeHdl:function (model) {
		OMV.Module.Storage.Greyhole.Admin.SMBPanel.superclass.cbSelectionChangeHdl.apply(this, arguments);
		// Process additional buttons
		this.toggleButtons();
	},

	toggleButtons:function () {
		var sm = this.getSelectionModel();
		var records = sm.getSelections();

		var tbarRemoveCtrl = this.getTopToolbar().findById(this.getId() + "-remove");

		if (records.length <= 0) {
			tbarRemoveCtrl.disable();
		} else {
			tbarRemoveCtrl.enable();
		}
	},

	cbAddBtnHdl:function () {
		var wnd = new OMV.Module.Storage.Greyhole.Admin.SMBDialog({
			uuid     :OMV.UUID_UNDEFINED,
			listeners:{
				submit:function () {
					this.doReload();
				},
				scope :this
			}
		});
		wnd.show();
	},

	cbEditBtnHdl:function () {
		var selModel = this.getSelectionModel();
		var record = selModel.getSelected();
		var wnd = new OMV.Module.Storage.Greyhole.Admin.SMBDialog({
			uuid     :record.get("uuid"),
			listeners:{
				submit:function () {
					this.doReload();
				},
				scope :this
			}
		});
		wnd.show();
	},

	cbRemoveShareBtnHdl:function () {
		var selModel = this.getSelectionModel();
		var record = selModel.getSelected();
		OMV.MessageBox.show({
			title  :_("Confirmation"),
			msg    :_("Do you want to remove the selected share from Greyhole?"),
			buttons:Ext.Msg.YESNO,
			fn     :function (answer) {
				if (answer == "no")
					return;
				var wnd = new OMV.ExecCmdDialog({
					title               :_("Removing share ..."),
					rpcService          :"Greyhole",
					rpcMethod           :"removeSMBShare",
					rpcArgs             :{ "uuid":record.get('uuid') },
					hideStart           :true,
					hideStop            :true,
					killCmdBeforeDestroy:false,
					listeners           :{
						finish   :function (wnd, response) {
							wnd.appendValue("\n" + _("Done ..."));
							wnd.setButtonDisabled("close", false);
						},
						exception:function (wnd, error) {
							OMV.MessageBox.error(null, error);
							wnd.setButtonDisabled("close", false);
						},
						close    :function () {
							this.doReload();
						},
						scope    :this
					}
				});
				wnd.setButtonDisabled("close", true);
				wnd.show();
				wnd.start();
			},
			scope  :this,
			icon   :Ext.Msg.QUESTION
		});
	},

	/** FSCK HANDLER */
	cbfsckBtnHdl:function () {
		var wnd = new OMV.Module.Storage.Greyhole.Admin.FSCKDialog({
			listeners:{
				success:function (wnd, path, email_report, dont_walk_metadata_store, find_orphaned_files, checksums, delete_rphaned_metadata) {
					this.dofsck(path, email_report, dont_walk_metadata_store, find_orphaned_files, checksums, delete_rphaned_metadata);
				},
				scope  :this
			},
			type     :"smb"
		});
		wnd.show();
	},
	dofsck      :function (path, email_report, dont_walk_metadata_store, find_orphaned_files, checksums, delete_rphaned_metadata) {
		OMV.Ajax.request(this.cbfsckLHdl, this, "Greyhole", "fsck",
						{
							path                    :String(path),
							email_report            :Boolean(email_report),
							checksums               :Boolean(checksums),
							dont_walk_metadata_store:Boolean(dont_walk_metadata_store),
							find_orphaned_files     :Boolean(find_orphaned_files),
							delete_orphaned_metadata:Boolean(delete_rphaned_metadata)
						}
		);
	},
	cbfsckLHdl  :function (id, response, error) {
		if (error !== null) {
			// Display error message
			OMV.MessageBox.error(null, error);
		} else {
			OMV.MessageBox.hide();
			this.doReload();
		}
	},
	/** /FSCK HANDLER */

	/** CANCEL FSCK HANDLER */
	cbunfsckBtnHdl:function () {
		this.dounfsck();
	},
	dounfsck      :function () {
		OMV.Ajax.request(this.cbunfsckLHdl, this, "Greyhole", "unfsck", []);
	},
	cbunfsckLHdl  :function (id, response, error) {
		if (error !== null) {
			// Display error message
			OMV.MessageBox.error(null, error);
		} else {
			OMV.MessageBox.hide();
			this.doReload();
		}
	}
	/** /CANCEL FSCK HANDLER */
});

OMV.NavigationPanelMgr.registerPanel("storage", "greyhole", {
	cls     :OMV.Module.Storage.Greyhole.Admin.SMBPanel,
	position:30,
	title   :_("SMB Shares")
});
